File: legacy/legacy-mahabharata-ios/Mahabharata/Model/DataClasses/Visual/Animations/TranslateAnim.java
package com.fulldome.mahabharata.model.visual.animation;

import com.fulldome.mahabharata.model.visual.Layer;

public class TranslateAnim extends LayerAnim {
	private int x;
	private int y;

	public TranslateAnim() {
	}

	public TranslateAnim(int x, int y) {
		this.x = x;
		this.y = y;
	}

	public int getX() {
		return x;
	}

	public int getY() {
		return y;
	}

	@Override
	public TranslateAnim interpolate(LayerAnim endAnim, float fraction) {
		TranslateAnim end = (TranslateAnim) endAnim;
		return new TranslateAnim(Layer.INT_EVALUATOR.evaluate(fraction, getX(), end.getX()), Layer.INT_EVALUATOR.evaluate(fraction, getY(), end.getY()));
	}

	@Override
	public void apply(Layer.ViewData data, int width, int height) {
		data.getMatrix().postTranslate(x, y);
	}
}

File: legacy/legacy-bhagavadgita.comics-ios-swift/Mahabharata/Model/DataClasses/Visual/Animations/TranslateAnim.swift
//
//  Comics.swift
//  Mahabharata
//
//  Created by Roman Developer on 11/28/17.
//  Copyright © 2017 Iron Water Studio. All rights reserved.
//

import UIKit

class TranslateAnim: Anim {
	let x: Int
	let y: Int
	
	private enum CodingKeys: String, CodingKey {
		case x
		case y
	}
	
	required init(from decoder: Decoder) throws {
		let container = try decoder.container(keyedBy: CodingKeys.self)
		self.x = (try? container.decode(Int.self, forKey: .x)) ?? 0
		self.y = (try? container.decode(Int.self, forKey: .y)) ?? 0
		
		try super.init(from: decoder)
	}
	
	init(x: Int, y: Int) {
		self.x = x
		self.y = y
		
		super.init()
	}
	
	override var description: String {
		return [
			super.description,
			"X: \(x)",
			"Y: \(y)"
			].joined(separator: ", ")
	}
	
	//MARK: - Overrides
	override func interpolate(endAnim: Anim, fraction: Double) -> Anim {
		let end = endAnim as! TranslateAnim
		//print("TranslateAnim interpolate. Fraction: \(fraction), start: \(self.y), end: \(end.y), interpolated y: \(LinearInterpolation.evaluate(fraction: fraction, startValue: self.y, endValue: end.y))")
		return TranslateAnim(x: LinearInterpolation.evaluate(fraction: fraction, startValue: self.x, endValue: end.x),
		                     y: LinearInterpolation.evaluate(fraction: fraction, startValue: self.y, endValue: end.y))
	}
	
	override func apply(to data: Layer.ViewData, width: Int, height: Int) {
		data.matrix = CATransform3DTranslate(data.matrix, CGFloat(self.x), CGFloat(self.y), 0)
	}
}

